src/functions/listFunctions.ts

```diff
@@ -0,0 +1,60 @@
+import Table = require("cli-table");
+import * as backend from "../deploy/functions/backend";
+import { logger } from "../logger";
+import { previews } from "../previews";
+import * as args from "../deploy/functions/args";
+
+/**
+ * Lists all functions in the Firebase project
+ * @param context the Context of the project
+ * @returns mapping that contains a list of functions under the 'functions' key
+ */
+export async function listFunctions(
+  context: args.Context
+): Promise<{ functions: backend.FunctionSpec[] }> {
+  const bkend = await backend.existingBackend(context, true);
+  const functionSpecs = previews.functionsv2
+    ? bkend.cloudFunctions
+    : bkend.cloudFunctions.filter((fn) => fn.platform === "gcfv1");
+  functionSpecs.sort((fn1, fn2) => {
+    if (fn1.entryPoint > fn2.entryPoint) {
+      return 1;
+    }
+    if (fn1.entryPoint < fn2.entryPoint) {
+      return -1;
+    }
+    return 0;
+  });
+  const table = previews.functionsv2
+    ? new Table({
+        head: ["Function", "Version", "Trigger", "Location", "Memory", "Runtime"],
+        style: { head: ["yellow"] },
+      })
+    : new Table({
+        head: ["Function", "Trigger", "Location", "Memory", "Runtime"],
+        style: { head: ["yellow"] },
+      });
+  for (const fnSpec of functionSpecs) {
+    const trigger = backend.isEventTrigger(fnSpec.trigger) ? fnSpec.trigger.eventType : "https";
+    if (previews.functionsv2) {
+      table.push([
+        fnSpec.entryPoint,
+        fnSpec.platform === "gcfv2" ? "v2" : "v1",
+        trigger,
+        fnSpec.region,
+        fnSpec.availableMemoryMb || "---",
+        fnSpec.runtime,
+      ]);
+    } else {
+      table.push([
+        fnSpec.entryPoint,
+        trigger,
+        fnSpec.region,
+        fnSpec.availableMemoryMb || "---",
+        fnSpec.runtime,
+      ]);
+    }
+  }
+  logger.info(table.toString());
```

Comment:
I feel like this entire logic could live in the `command/functions-list.ts` instead of as a its own module.

I think we are trying to keep the logic under `command/functions-list.ts` as light as possible, but most of the logic here seems to be about cosmetics of the output, which (I think) should live in the command side since comestic codes are less shareable.

At the same time, if all of the cosmetic code is pull out of here, then we are basically left with `backend.existingBackend` - maybe this function can just be responsible for returning the right set of functions given user's preview flag?

@bkendall recently shared a great thought on Firebase CLI chat room what he thinks should and should not happen in the command code - might want to ping him for opinion!

Reply:
I struggled a bit on where to put this logic too. I read and re-read @bkendall's comment and having all the logic in this function does make the `Command` very flat. But on the other hand, if the `Command`'s job is to format output for the user, then we should move the `Table` stuff there and just leave the API call here (which kind of makes this function useless if it's just calling an existing api & sorting the results). Also, the pattern that I used here already exists in the project - `ext:list` uses this technique pretty much exactly. 

I'm curious to see what others think. CC @inlined @bkendall

Reply:
While I would not have made the same decision as @bkendall, I want to respect his experience and authority. If I can paraphrase everyone's concerns, this seems like an MVC codebase. We don't mind mixing models and controllers (e.g. most of backend.ts and existingBackend()), but we're not sure views belong in the same place. I would love some thoughts on how to cleanly separate the view codebases from the model codebases, but this is a fine first start.

Reply:
OTOH, unit tests are only verifying that the sort function works. If that's all we're going to test, then we can put the sort function in backend.ts per my other comment and move this untested code to the command.

Reply:
(I like your MVC codebase comparison. The idea that I usually base my thoughts off of are "functional core, imperative shell". Short version: the less the `Command` has to do with the actual core of getting the data and making decisions, the better. `Command`s shouldn't know how to make API calls, and the core (like `listFunctions` shouldn't know how to write tables to stdout.)